src/app/AiPlayer.ts

```diff
@@ -0,0 +1,195 @@
+  /**
+   * Makes a move for AI Player
+   * @param gameController for the game that the player is participating in
+   */
+  makeAMove(gameController: GameController): void {
+    this.trade(gameController);
+    if (gameController.theGameProcessor.checkWin()) {
+      gameController.checkIfGameIsWon();
+      return;
+    }
+    gameController.breed();
+    if (gameController.theGameProcessor.checkWin()) {
+      gameController.checkIfGameIsWon();
+      return;
+    }
+  }
+  private trade(gameController: GameController): void {
+    //Try to win the game by trading one to many if possible
+    const oneToManyTrades: [AnimalNames, Offer[]][] = [
+      [
+        AnimalNames.HORSE,
+        [
+          [AnimalNames.COW, 1],
+          [AnimalNames.PIG, 1],
+          [AnimalNames.SHEEP, 3],
+          [AnimalNames.RABBIT, 6],
+        ],
+      ],
+      [
+        AnimalNames.COW,
+        [
+          [AnimalNames.PIG, 1],
+          [AnimalNames.SHEEP, 3],
+          [AnimalNames.RABBIT, 6],
+        ],
+      ],
+      [
+        AnimalNames.PIG,
+        [
+          [AnimalNames.SHEEP, 1],
+          [AnimalNames.RABBIT, 6],
+        ],
+      ],
+      [AnimalNames.SHEEP, [[AnimalNames.RABBIT, 6]]],
+    ];
+    for (const [animalToSell, target] of oneToManyTrades) {
+      if (this.herd.getAnimalNumber(animalToSell) > 1) {
+        const offer: Offer[] = [[animalToSell, 1]];
+        if (
+          gameController.theGame.theTrade.processOffer(
+            offer,
+            this,
+            target,
+          )
+        ) {
+          return;
+        }
+      } else if (this.herd.getAnimalNumber(animalToSell) === 0) {
+        break;
+      }
+    }
+
+    //Try to win by trading many to one (possible if only one kind of animal is missing)
+    const animalsInOrder = [
+      AnimalNames.HORSE,
+      AnimalNames.COW,
+      AnimalNames.PIG,
+      AnimalNames.SHEEP,
+      AnimalNames.RABBIT,
+    ];
+
+    const missingAnimals = animalsInOrder.filter(
+      (animal) => this.herd.getAnimalNumber(animal) === 0,
+    );
+
+    if (missingAnimals.length === 1) {
+      const offer = this.animalsForTrade();
+      const target: Offer[] = [[missingAnimals[0], 1]];
+      if (
+        gameController.theGame.theTrade.processOffer(
+          offer,
+          this,
+          target,
+        )
+      ) {
+        return;
+      }
+    }
+    //Cannot win game by trade. Check if buying dogs is worth it
+    if (this.buyABigDog()) {
+      if (
+        gameController.theGame.theTrade.processOffer(
+          [[AnimalNames.PIG, 3]],
+          this,
+          [[AnimalNames.BIG_DOG, 1]],
+        )
+      ) {
+        return;
+      } else if (
+        gameController.theGame.theTrade.processOffer(
+          [[AnimalNames.COW, 1]],
+          this,
+          [[AnimalNames.BIG_DOG, 1]],
+        )
+      ) {
+        return;
+      }
+    }
+
+    if (this.buyASmallDog()) {
+      const offer: Offer[] = [[AnimalNames.RABBIT, 6]];
```

Comment:
See if this could be also automated by counting the values of animals in the herd (every animal that the ai has above 1) until it reaches the value of the small dog (which should also come from config)

Reply:
Same as above.